noir-projects/noir-protocol-circuits/crates/reset-kernel-lib/src/reset/key_validation_hint.nr

```diff
@@ -28,18 +26,19 @@ impl Eq for KeyValidationHint {
 }
 
 pub fn reset_key_validation_requests<N>(
-    key_validation_requests: [ScopedKeyValidationRequest; MAX_KEY_VALIDATION_REQUESTS_PER_TX],
+    key_validation_requests: [ScopedKeyValidationRequestAndGenerator; MAX_KEY_VALIDATION_REQUESTS_PER_TX],
     key_validation_hints: [KeyValidationHint; N]
-) -> BoundedVec<ScopedKeyValidationRequest, MAX_KEY_VALIDATION_REQUESTS_PER_TX> {
+) -> BoundedVec<ScopedKeyValidationRequestAndGenerator, MAX_KEY_VALIDATION_REQUESTS_PER_TX> {
     let mut should_propagate = key_validation_requests.map(|req| !is_empty(req));
     for i in 0..N {
         let hint = key_validation_hints[i];
         if !is_empty(hint) {
```

Comment:
Hints should be helpful for checking but not deciding if we are checking. As I understand they are based on the oracles, so If I have a malicious oracle I could just set no hints and then skip all the validation? 

I don't think that entering the if should be based on constrained values instead of the user controlled hints, e.g., use `key_validation_requests` instead.

Reply:
Super great catch! This is a vulnerability.

Will fix it by checking the validation request instead and put there a note.

Reply:
This is by design. Now we have reset circuits that can be run between any two private circuits, it's not necessary to reset all data in one go. Especially for transient note hashes and nullifiers: if the array is about to full, we can run a reset circuit to remove the transient data, so that we can keep adding more notes and nullifiers in the following calls. But we can't reset all of them all the time. A transient note might be read in a nested call so it will have to be kept until the read request is verified. (Related to your questions above, that's why sometimes we need to propagate requests to the public inputs.)

And that's why we are using hints to decide which data to verify and which to propagate to the public inputs. However, all the requests must be verified and reset eventually. The tail circuit will check that all of those arrays are empty.

For key validations, it's not like transient data that might need to be preserved for later iterations. But we still prefer using hints instead of requests to loop through the data because then the circuit size is defined by the size of the hints array. We can then set a really big number for the MAX_XXX_PER_TX, and if there aren't many XXX present, we only need to run a small reset circuit that takes an array of 10 hints and only execute the verification code 10 times.

Reply:
Cool! So it's not a vulnerability after all. Will close the issue and add there a comment. Thanks for the explanation @LeilaWang